### PR TITLE
NEXUS-6048: Index based release remover

### PR DESCRIPTION
**Pull request by cstamas**
Changes: task modification by adding "use index" checkbox, pulling out backends with existing unmodified "walker" implementation, and added new "index" based one that is actually a "hybrid", due to some limitations of MI.

Issue
https://issues.sonatype.org/browse/NEXUS-6048

CI
http://bamboo.s/browse/NX-OSSF600

### CODE REVIEW

File: components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/ReleaseRemovalRequest.java [hunk 1]
```diff
@@ -46,6 +56,10 @@ public int getNumberOfVersionsToKeep() {
     return numberOfVersionsToKeep;
   }
 
+  public boolean isIndexBackend() {
+    return indexBackend;
+  }
+
   public String getTargetId() {
```
Comment (mcculls) [inline components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/ReleaseRemovalRequest.java]: Perhaps mark return value as `@Nullable` to be consistent with constructor?

File: components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/ReleaseRemovalRequest.java [hunk 2]
```diff
@@ -25,16 +27,24 @@
...
   /**
    * @param repositoryId           the repository to target
    * @param numberOfVersionsToKeep the number of released versions to keep for Group/Artifact in the repository
+   * @param indexBackend           use index instead of walker
    * @param targetId               (optional) Repository Target id to be applied
    */
-  public ReleaseRemovalRequest(final String repositoryId, final int numberOfVersionsToKeep, final String targetId) {
+  public ReleaseRemovalRequest(final String repositoryId,
+                               final int numberOfVersionsToKeep,
+                               final boolean indexBackend,
```
Comment (mcculls) [inline components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/ReleaseRemovalRequest.java]: Minor: imho `useIndex` is a more descriptive name here

File: components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/WalkerReleaseRemoverBackend.java [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+                             final ReleaseRemovalResult result,
+                             final MavenRepository repository,
+                             final @Nullable Target repositoryTarget)
+      throws IOException
+  {
+    TaskUtil.checkInterruption();
+
+    log.debug(
+        "Collecting deletable releases on repository '" + repository.getId() + "' from storage directory "
+            + repository.getLocalUrl());
```
Comment (mcculls) [inline components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/WalkerReleaseRemoverBackend.java]: Change to use `{}` markers

File: components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/WalkerReleaseRemoverBackend.java [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+        new HashMap<Gav, Map<Version, List<StorageFileItem>>>();
+
+    private final ReleaseRemovalResult result;
+
+    private final Target repositoryTarget;
+
+    private int deletedFiles = 0;
+
+    private ReleaseRemovalWalkerProcessor(final MavenRepository repository,
+                                          final ReleaseRemovalRequest request, final ReleaseRemovalResult result,
```
Comment (mcculls) [inline components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/WalkerReleaseRemoverBackend.java]: odd formatting?

File: components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/descriptors/ReleaseRemovalTaskDescriptor.java [hunk 5]
```diff
@@ -48,7 +51,9 @@
           NUMBER_OF_VERSIONS_TO_KEEP_FIELD_ID, "Number to keep", "The number of versions for each GA to keep",
           FormField.MANDATORY),
       new RepoTargetComboFormField(REPOSITORY_TARGET_FIELD_ID, "Repository Target",
-          "Select a repository target to apply", FormField.OPTIONAL)
+          "Select a repository target to apply", FormField.OPTIONAL),
+      new CheckboxFormField(INDEX_BACKEND, "Use index",
+          "Removal should happen by crawling the repository (unchecked) or by using Maven Index?", FormField.MANDATORY)
```
Comment (mcculls) [inline components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/descriptors/ReleaseRemovalTaskDescriptor.java]: Could reword to: `Use Maven Index to walk the repository? Defaults to crawling if unchecked`

File: components/nexus-core/src/test/java/org/sonatype/nexus/maven/tasks/DefaultReleaseRemoverIT.java [hunk 6]
```diff
@@ -43,6 +48,8 @@
 public class DefaultReleaseRemoverIT
     extends AbstractMavenRepoContentTests
 {
+  // we have no index backend available here!
+  public static final boolean indexBackend = false;
```
Comment (mcculls) [inline components/nexus-core/src/test/java/org/sonatype/nexus/maven/tasks/DefaultReleaseRemoverIT.java]: Formatting: upper-case for constants

File: plugins/indexer/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/index/releaseremover/IndexReleaseRemoverBackend.java [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+import org.apache.lucene.search.Query;
+import org.apache.maven.index.ArtifactInfo;
+import org.apache.maven.index.IteratorSearchResponse;
+import org.apache.maven.index.MAVEN;
+import org.apache.maven.index.context.IndexingContext;
+import org.apache.maven.index.expr.SourcedSearchExpression;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
```
Comment (mcculls) [inline plugins/indexer/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/index/releaseremover/IndexReleaseRemoverBackend.java]: Minor: describe what this class does

**Comment by mcculls** [comment]
+1

File: components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/ReleaseRemovalRequest.java (hunk 2 above)
Comment (kellyrob99) [inline components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/ReleaseRemovalRequest.java]: Agreed

File: components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/descriptors/ReleaseRemovalTaskDescriptor.java (hunk 5 above)
Comment (kellyrob99) [inline components/nexus-core/src/main/java/org/sonatype/nexus/maven/tasks/descriptors/ReleaseRemovalTaskDescriptor.java]: Agreed, makes more sense to describe what happens when checked first

File: plugins/indexer/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/index/releaseremover/IndexReleaseRemoverBackend.java [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+              final String gavPath = repository.getGavCalculator()
+                  .gavToPath(new Gav(groupId, artifactId, version.toString()));
+              final String vDirectory = PathUtils.getParentPath(gavPath);
+              try {
+                int deleted = mayDeleteVersion(target, repository, vDirectory);
+                log.debug("Deleted {} files from {}:{}:{} in {}", deleted, groupId, artifactId, version, repository);
+                deletedFiles += deleted;
+              }
+              catch (Exception e) {
+                log.warn("Could not delete {}:{}:{} in {}, skipping it", groupId, artifactId, version, repository, e);
```
Comment (kellyrob99) [inline plugins/indexer/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/index/releaseremover/IndexReleaseRemoverBackend.java]: should this affect the 'success' status?

File: plugins/indexer/nexus-indexer-lucene-plugin/src/test/java/org/sonatype/nexus/index/releaseremover/IndexReleaseRemoverIT.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+import static org.junit.Assert.assertThat;
+
+/**
+ * @since 2.11.3
+ */
+public class IndexReleaseRemoverIT
+    extends AbstractIndexerManagerTest
+{
+  // we test the index only
+  public static final boolean indexBackend = true;
```
Comment (kellyrob99) [inline plugins/indexer/nexus-indexer-lucene-plugin/src/test/java/org/sonatype/nexus/index/releaseremover/IndexReleaseRemoverIT.java]: private and UPPER?

**Comment by kellyrob99** [comment]
+1

File: plugins/indexer/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/index/releaseremover/IndexReleaseRemoverBackend.java (hunk 8 above)
Comment (cstamas) [inline plugins/indexer/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/index/releaseremover/IndexReleaseRemoverBackend.java]: Ack, it should

**Comment by cstamas** [comment]
http://bamboo.s/browse/NX-OSSF600-2 :+1:

**Comment by cstamas** [comment]
Merged as 92b44a50a3d3b3d30cc6ee95ad9555f5b01681d8

**Pull request closed without merging**